### ISSUE TITLE
--cors not working

### ISSUE THREAD

**Issue by chovy** [http-party/http-server#394]
`XMLHttpRequest cannot load http://127.0.0.1:8081/feeds.json. Request header field Authorization is not allowed by Access-Control-Allow-Headers in preflight response`

```
$ http-server --cors
Starting up http-server, serving ./
Available on:
  http://127.0.0.1:8081
  http://192.168.1.3:8081
```

Here are the headers:

```
Response Headers
access-control-allow-headers:Origin, X-Requested-With, Content-Type, Accept, Range
access-control-allow-origin:*
Connection:keep-alive
Date:Sun, 03 Sep 2017 07:51:28 GMT

Request Headers
view source
Accept:*/*
Accept-Encoding:gzip, deflate, br
Accept-Language:en-US,en;q=0.8
Access-Control-Request-Headers:authorization
Access-Control-Request-Method:GET
```

This is the OPTIONS request in chrome that is failing.

**Comment by jasonxia23** [comment]
`--cors` works fine for me.
Check your request headers whether there are headers not in `Origin, X-Requested-With, Content-Type, Accept, Range`

**Comment by chovy** [comment]
I had to pass `--cors=Authorization`

**Comment by TeodorKolev** [comment]
`--cors` option do nothing for me. I have tried `http-server --cors` and  `http-server --cors='*'`
No success

**Comment by Kazanz** [comment]
Same problem here.

**Comment by BigBlueHat** [comment]
@TeodorKolev @ansermino @Kazanz are you all simply using `--cors` (without a value or with `'*'` in @TeodorKolev's case)?

The test for this feature is here:
https://github.com/indexzero/http-server/blob/413fe72f8da11c7c5746d16e4a7876450f2d9d8c/test/http-server-test.js#L128-L156

Please extend it to test your scenario/usage expectation and lets see what we can figure out! 😃

**Comment by Kazanz** [comment]
@BigBlueHat After fiddling around turns out CORS was working in Firefox but not in chrome, because of the capitalization of the header.

**Comment by PhilTheAir** [comment]
Any updates on this one? Thanks.

**Comment by BigBlueHat** [comment]
> @BigBlueHat After fiddling around turns out CORS was working in Firefox but not in chrome, because of the capitalization of the header.

Capitalization of the CORS `Allow-*` headers? or of the header passed in via `--cors=`? [HTTP headers are not case sensitive](https://tools.ietf.org/html/rfc7230#section-3.2), so it would surprise me if Chrome were failing for that reason.

We'll need to narrow down the scenario and test it before we change the code for this feature. Thanks!

**Comment by BigBlueHat** [comment]
> This one works for me: http-server {folder} --cors=http://localhost:3000/

@PhilTheAir the `--cors` takes a list of HTTP Headers. It's currently not possible to narrow the `Access-Control-Allow-Origin` header to something less than `*`. If you need that, please file another issue for it--as it's likely useful to others! See https://github.com/indexzero/http-server/blob/master/lib/http-server.js#L69-L79 for codez.

**Comment by danieladams456** [comment]
This is potentially unrelated to some of the more specific comments, but in general if you're only testing in browser, the default caching of one hour can cause the non-cors response to be cached and fail to get the new cors response.

**Comment by J-Rojas** [comment]
I'm having a problem with reading headers within an XHR request in Chrome. 'Access-Control-Expose-Headers' needs to be supported as well.

https://github.com/mozilla/pdf.js/issues/3150#issuecomment-17582371  
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers

**Comment by Mathes2665** [comment]
I had the same problem. I solved it by installing cors global `npm install cors -g`.

**Comment by bhagwantg** [comment]
--cors without value works, probably browser cache issue, hit ctrl + f5

**Comment by yinleiCoder** [comment]
It may be the cause of browser caching. 
I solved the reason why cros did not work, and forced to refresh the browser by ctrl+F5

**Comment by piyush94** [comment]
--cors option adds 'Access-Control-Allow-*' headers in lowercase
![image](https://user-images.githubusercontent.com/11614204/101757799-94a65e00-3afd-11eb-8afb-2d51cad7afe3.png)

Swagger UI is giving error due to this.

**Comment by ryanbuckleyca** [comment]
same issue here. tried `npx http-server --cors='*'`,` --cors`, `--cors=Authorization`, tried hard-refreshing the browser and clearing the cache, tried in multiple browsers...still getting the following issue.

> Access to fetch at 'http://www.theurliwanttofetch.com' from origin 'http://127.0.0.1:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Comment by malibio** [comment]
I'm on the same scenario here...what ended up working for you, @ryanbuckleyca ?

> same issue here. tried `npx http-server --cors='*'`,` --cors`, `--cors=Authorization`, tried hard-refreshing the browser and clearing the cache, tried in multiple browsers...still getting the following issue.
> 
> > Access to fetch at 'http://www.theurliwanttofetch.com' from origin 'http://127.0.0.1:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Comment by akshayr-tavisca** [comment]
Hello, facing similar cors issue,  when trying to load some external file within the application, 
`hs . --cors -p 8080` any updates on above issue, it works in few cases and fails in few and I am not sure why.

**Comment by LetrixZ** [comment]
```--cors``` works for me but Ctrl+F5 nor clearing site's data fixes the cache issue. Only on incognito or in another browser.

**Comment by thornjad** [comment]
Is anyone having trouble with this on v13 and cache resetting does _not_ fix the problem?

**Comment by abitgen** [comment]
`--cors` works! Just restarted the node http-server process. Using node v14.15.0 & http-server 13.0.0

**Comment by thornjad** [comment]
:tada: I'll close this as fixed in v13 then, please reopen if it appears broken again _and_ resetting cache does not fix it

**Comment by thornjad** [comment]
@BeyYT what is everything you did and tried?

**Comment by yuchinchenTW** [comment]
I tried http-server --cors=* -p8110 -c-1 but the browser shows" has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled."

**Comment by BeyYT** [comment]
> @BeyYT what is everything you did and tried?

Every option, CORS header, even * and it still doesn’t work.

**Comment by abitgen** [comment]
@BeyYT a sample repo link, with the http-server version you are using, might help.

**Comment by BeyYT** [comment]
> @BeyYT a sample repo link, with the http-server version you are using, might help.

the [sample repo](https://github.com/BeyYT/beatshape-web) has been on my github for a long time now. aswell the http-server version is the latest version there is.
